Revise math::factorize()

bertwesarg opened the pull request.

bertwesarg committed:
doc: 4 is not prime

bertwesarg committed:
math::factors: Reuse math::factorize()

bertwesarg committed:
math::factorize: Avoid condition in loops

bertwesarg committed:
math::factorize: Drop special case for 2

bertwesarg committed:
math::factorize: Avoid sqrt and use n for the upper bound

devreal commented on dash/include/dash/internal/Math.h:
```diff
@@ -161,44 +143,16 @@ std::map<Integer, int> factorize(Integer n)
-    }
+  const auto factorization = factorize(n);
+  for(auto it = factorization.begin(); it != factorization.end(); ++it) {
+    primes.insert(it->first);
```
Small further improvement: `primes.insert(factorization.begin(), factorization.end())`

bertwesarg replied:
are you sure? `factorization` is a `std::map` where the iterators gives you `std::pair`s

devreal replied:
Ahh, you're right of course. Never mind then :)

devreal commented on dash/include/dash/internal/Math.h:
```diff
@@ -107,7 +107,7 @@ double sigma(const Iter& begin, const Iter& end)
+ *   // returns map { (2,5), (7,2) }
  * \endcode
  */
 template <typename Integer>
```
Add a `static_assert` that `Integer` is indeed an integral type?

bertwesarg replied:
It should also be `unsigned`? And the exponents too?

devreal replied:
IMO it's better to add a runtime assert for positive numbers to avoid casting a negative number to `unsigned`. The factors should be unsigned though, yes.

bertwesarg replied:
`static_assert` is compile time

bertwesarg replied:
I think I got it:

* [x] add `static_assert` to check that `Integer` is an integral type
* [x] add `assert` that `n => 0`
* [x] change mapped type from `int` to `unsigned`

correct?

devreal replied:
That's what I meant, yes :)

devreal reviewed (approved):
Looks good to me. I wanted to ask for tests but realized that we have `AutobalanceTest.Factorize` :) Just two minor comments

devreal reviewed (changes requested):
Tests fail. Example: 

```
[  ERROR   ] [UNIT 0] in /root/project/dash/test/pattern/TilePatternTest.cc:56
Expected equality of these values:
  dash::size()
    Which is: 2
  teamspec_2d.size()
    Which is: 1
[  ERROR   ] [UNIT 1] in /root/project/dash/test/pattern/TilePatternTest.cc:56
Expected equality of these values:
  dash::size()
    Which is: 2
  teamspec_2d.size()
    Which is: 1
[  FAILED  ] TilePatternTest.Tile2DimTeam2Dim
[  ERROR   ] [UNIT 0] in /root/project/dash/test/TestPrinter.h:149
Failed
Testcase failed at least on one unit
```

bertwesarg committed:
test: Drop unused GTest forward declarations

bertwesarg committed:
gtest: Link against full external gtest library by path instead of by name

bertwesarg committed:
test: Use descriptive names in the `EQAsserter` class

Before:

```
Expected equality of these values:
  _e
    Which is: 10
  _a
    Which is: 0
```

After:

```
Expected equality of these values:
  expected
    Which is: 10
  actual
    Which is: 0
```

bertwesarg commented:
Can someone triggers the CI again please.

bertwesarg committed:
math::factorize: Add type and value checks

bertwesarg commented:
> Tests fail. Example:
> 
> ```
> [  ERROR   ] [UNIT 0] in /root/project/dash/test/pattern/TilePatternTest.cc:56
> Expected equality of these values:
>   dash::size()
>     Which is: 2
>   teamspec_2d.size()
>     Which is: 1
> [  ERROR   ] [UNIT 1] in /root/project/dash/test/pattern/TilePatternTest.cc:56
> Expected equality of these values:
>   dash::size()
>     Which is: 2
>   teamspec_2d.size()
>     Which is: 1
> [  FAILED  ] TilePatternTest.Tile2DimTeam2Dim
> [  ERROR   ] [UNIT 0] in /root/project/dash/test/TestPrinter.h:149
> Failed
> Testcase failed at least on one unit
> ```

CI is now green.

devreal reviewed (approved):
Looks good to me now :)
